### PR TITLE
Refactor JSONExtract functions and support more types to reuse its code in new JSON type

### PR DESCRIPTION
**Pull request by Avogar**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Refactor JSONExtract functions, support more types including experimental Dynamic type.

This refactoring will allow to reuse JSONExtract functions code in new JSON type.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by Avogar** [comment]
I moved code from `FunctionsJSON.h` into 2 different files, first - `JSONExtractTree.h/cpp` with the code of building the tree for later insertion, and second - `FunctionsJSON.cpp` with the code of the functions (no need to do it in the header).

**Comment by robot-clickhouse-ci-1** [comment]
*This is an automated comment for commit ea3b0e735de285db89cb36e2782db88c6d403ee2 with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/66046/ea3b0e735de285db89cb36e2782db88c6d403ee2/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Style check</td><td>Runs a set of checks to keep the code style clean. If some of tests failed, see the related log from the report</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/66046/ea3b0e735de285db89cb36e2782db88c6d403ee2/style_check.html">❌ failure</a></td></tr>
<tbody>
</table>

File: src/Functions/FunctionsJSON.cpp [hunk 1]
```diff
@@ -1,10 +1,1060 @@
...
+template <typename Name, template<typename> typename Impl>
+class ExecutableFunctionJSON : public IExecutableFunction
+{
+
+public:
+    explicit ExecutableFunctionJSON(const NullPresence & null_presence_, bool allow_simdjson_, const DataTypePtr & json_return_type_, const FormatSettings & format_settings_)
+        : null_presence(null_presence_), allow_simdjson(allow_simdjson_), json_return_type(json_return_type_), format_settings(format_settings_)
+    {
+        format_settings.json.escape_forward_slashes = false;
+        format_settings.null_as_default = false;
```
Comment (antaljanosbenjamin) [inline src/Functions/FunctionsJSON.cpp:358]: I am not sure why are these changes necessary. Could you please explain it?

File: src/Formats/JSONExtractTree.cpp [hunk 2]
```diff
@@ -0,0 +1,1574 @@
...
+        auto it = object.begin();
+        for (; it != object.end(); ++it)
+        {
+            auto pair = *it;
+
+            /// Insert key
+            key_col.insertData(pair.first.data(), pair.first.size());
+
+            /// Insert value
+            if (!value-> insertResultToColumn(value_col, pair.second, insert_settings, format_settings, error))
```
Comment (antaljanosbenjamin) [inline src/Formats/JSONExtractTree.cpp:1201]:
```suggestion
            if (!value->insertResultToColumn(value_col, pair.second, insert_settings, format_settings, error))
```

Maybe could you please format the whole file with `clang-format?`

File: src/Formats/JSONExtractTree.h [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+#include <Formats/FormatSettings.h>
+
+
+namespace DB
+{
+
+struct JSONExtractInsertSettings
+{
+    bool convert_bool_to_integer = true;
+    bool insert_default_on_invalid_elements_in_complex_types = false;
```
Comment (antaljanosbenjamin) [inline src/Formats/JSONExtractTree.h:13]: Could you please provide a brief description for why are these new settings are needed? Do we want to handle these differently in json functions and in the new JSON type?

File: src/Formats/JSONExtractTree.cpp [hunk 4]
```diff
@@ -0,0 +1,1566 @@
...
+                if (auto type = tryInferDateOrDateTimeFromString(data, format_settings))
+                    return type;
+
+                if (format_settings.json.try_infer_numbers_from_strings)
+                {
+                    bool is_negative = false;
+                    if (auto type = tryInferJSONNumberFromString(data, format_settings, &json_inference_info))
+                    {
+                        json_inference_info.numbers_parsed_from_json_strings.insert(type.get());
+                        if (is_negative)
```
[the remark this answers is not in the document]
Comment (Avogar) [reply src/Formats/JSONExtractTree.cpp:1355]: Good notice, I forgot to remove it from here when I added check for negative numbers in SchemaInferenceUtils. Will remove it

File: src/Formats/JSONExtractTree.cpp (hunk 2 above)
Comment (Avogar) [reply src/Formats/JSONExtractTree.cpp:1201]: Sure

File: src/Formats/JSONExtractTree.h (hunk 3 above)
Comment (Avogar) [reply src/Formats/JSONExtractTree.h:13]: These settings are needed to keep the current behaviour of JSONExtract functions.

`convert_bool_to_integer` is set to `false` for functions `JSONExtractUInt64/JSONExtractInt64/JSONExtractFloat64` so we extract only real numbers (don't know the motivation, but we need to keep it as it was).

`insert_default_on_invalid_elements_in_complex_types` is set to true in all `JSONExtract` functions. It allows to do this:
```
:) select JSONExtract('{"a" : [1, "hello", 2]}', 'a', 'Array(Int32)')

SELECT JSONExtract('{"a" : [1, "hello", 2]}', 'a', 'Array(Int32)')

Query id: c401e109-634d-4495-a1bf-831029fac874

   ┌─JSONExtract('{"a" : [1, "hello", 2]}', 'a', 'Array(Int32)')─┐
1. │ [1,0,2]                                                     │
   └─────────────────────────────────────────────────────────────┘
```

So when we can parse only some values from an array/map, we keep them and insert default on the invalid values. Again, it how it was implemented in `JSONExtract`, we had to keep it. But we don't need it in new JSON type

File: src/Formats/JSONExtractTree.cpp (hunk 2 above)
Comment (Avogar) [reply src/Formats/JSONExtractTree.cpp:1201]: `value-> insertResultToColumn`
Seems like I made a mistake during some find and replace :)

**Review by antaljanosbenjamin — APPROVED** [review]
